File: easy/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        stack = []
        for aChar in s:
            if aChar == '(':
                stack.append(')')
            elif aChar == '{':
                stack.append('}')
            elif aChar == '[':
                stack.append(']')
            else:
                if not stack or stack.pop() != aChar:
                    return False

        return False if stack else True


# @lc code=end

File: easy/20.有效的括号.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  if (s.length % 2 === 1) {
    return false;
  }
  const stack = [];
  for (let aChar of s) {
    switch (aChar) {
      case '(':
        stack.push(')');
        break;
      case '[':
        stack.push(']');
        break;
      case '{':
        stack.push('}');
        break;
      default:
        if (stack.pop() !== aChar) {
          return false;
        }
    }
  }
  return stack.length === 0;
};
// @lc code=end

